FAIMS/FAIMS3
# feat: implement token refresh and on-demand refresh in app upon created new survey

# feat: implement token refresh and on-demand refresh in app upon created new survey

## JIRA Ticket

https://jira.csiro.au/browse/BSS-438
https://jira.csiro.au/browse/BSS-439

## Description

Implements a backend API stateful token refresh workflow. 

Adds logic in front end to refresh when a new notebook is created so that the updated token has the new roles, enabling the user to see the updated list. 

## Proposed Changes

This PR touched a **lot** of files as I was chasing through some unnecessarily complex/redundant handling of the tokens in the front end. Apologies for the huge diff. I made some improvements along the way, notably

- fixed initialisation bug which was causing listings to intermittently not appear on fresh load until refreshed
- added the parsed token contents into the local auth db so that you don't need to repeatedly parse the token contents
- new approach for the new Auth DB which makes use of native javascript code which is serialised into the DB upon init 
- removed a lot of `useState(token)` - this is unnecessary as the token is available from the local auth db

## TODOs before merge

- unit tests for API 
- developer documentation on refresh logic

## How to Test

Clear your local storage, logout, login. Your local storage will now have a refresh token. 

Then setup a user which has the create notebooks permission but not admin. Try creating a new survey. The survey list should be updated to reflect the new notebook, showing the token has the correct permissions/roles. 


## Checklist
